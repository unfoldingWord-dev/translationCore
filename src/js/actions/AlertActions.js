import types from './ActionTypes';

export const closeAlert = (id) => ({
  type: types.CLOSE_ALERT,
  id,
});

/**
 * Opens a new alert.
 * @param {string} id - the alert id
 * @param {*} message - the message to display in the alert
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @return {*}
 */
export const openAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  confirmText = null,
  cancelText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
});

/**
 * Opens a new alert that can be ignored.
 * @param {string} id - the alert id
 * @param {*} message
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {func} onIgnore - custom ignore handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @param {string} ignoreText - the ignore checkbox text
 * @return {*}
 */
export const openIgnorableAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  onIgnore = null,
  confirmText = null,
  cancelText = null,
  ignoreText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
  onIgnore: typeof onIgnore === 'function' ? onIgnore : () => {},
  ignoreText,
});

/**
 * Ignores a specific alert for the rest of the session
 * @param {string} id - the alert id to ignore
 * @param {boolean} ignore - ignores or enables the alert
 * @return {*}
 */
export const ignoreAlert = (id, ignore = true) => ({
  type: types.IGNORE_ALERT,
  id,
  ignore,
});
