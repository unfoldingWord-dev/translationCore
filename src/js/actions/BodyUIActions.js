import {
  getHomeScreenStep,
  getActiveHomeScreenSteps
} from '../selectors';
import types from './ActionTypes';

/**
 * @description toggles the home view based on param.
 * @param {boolean} boolean - true or false either shows or hides it.
 */
export const toggleHomeView = boolean => ({
  type: types.TOGGLE_HOME_VIEW,
  boolean
});

export const toggleWelcomeSplash = () => ({
  type: types.TOGGLE_WELCOME_SPLASH
});

export const goToNextStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex + 1));
  });
};

export const goToPrevStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex - 1));
  });
};

/**
 * Goes to specified step
 * @param {number} stepNumber - Number of step to go to in home stepper
 */
export const goToStep = stepNumber => {
  return ((dispatch, getState) => {
    if (stepNumber >= 0 && stepNumber <= 3) {
      const activeSteps = getActiveHomeScreenSteps(getState());
      if (!activeSteps[stepNumber]) return;
      dispatch({
        type: types.GO_TO_STEP,
        stepIndex: stepNumber
      });
    } else if (stepNumber < 0) {
      console.error('The min number of steps is 0. (0-3)');
    } else {
      console.error('The max number of steps is 3. (0-3)');
    }
  });
};

export const toggleProjectsFAB = () => ({
  type: types.TOGGLE_PROJECTS_FAB
});

export const openOnlineImportModal = () => ({
  type: types.OPEN_ONLINE_IMPORT_MODAL
});

export const closeOnlineImportModal = () => {
  return {
    type: types.CLOSE_ONLINE_IMPORT_MODAL
  };
};

export const openLicenseModal = () => ({
  type: types.OPEN_LICENSE_MODAL
});

export const closeLicenseModal = () => ({
  type: types.CLOSE_LICENSE_MODAL
});

/**
 * @description show or not show dimmed screen.
 * @param {bool} enable dims the screen if enabled otherwise removes the dim
 */
export const dimScreen = (enable=true) => ({
  type: types.SHOW_DIMMED_SCREEN,
  bool: enable
});
