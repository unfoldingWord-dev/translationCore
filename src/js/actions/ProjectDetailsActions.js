import React from 'react';
import consts from './ActionTypes';
import path from 'path-extra';
import fs from 'fs-extra';
import ospath from 'ospath';
import * as bibleHelpers from '../helpers/bibleHelpers';
import * as ProjectDetailsHelpers from '../helpers/ProjectDetailsHelpers';
import * as AlertModalActions from "./AlertModalActions";
import {getTranslate} from "../selectors";
import {cancelProjectValidationStepper} from "./ProjectImportStepperActions";
import * as ProjectOverwriteHelpers from "../helpers/ProjectOverwriteHelpers";
import * as ProjectImportFilesystemActions from "./Import/ProjectImportFilesystemActions";
// constants
const INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

/**
 * @description sets the project save location in the projectDetailReducer.
 * @param {String} pathLocation - project save location and/or directory.
 * @return {object} action object.
 */
export const setSaveLocation = pathLocation => {
  return((dispatch) => {
    dispatch({
      type: consts.SET_SAVE_PATH_LOCATION,
      pathLocation
    });
  });
};

export const resetProjectDetail = () => {
  return {
    type: consts.RESET_PROJECT_DETAIL
  };
};

export function setProjectToolGL(toolName, selectedGL) {
  return (dispatch) => {
    if(typeof toolName !== 'string') {
      return Promise.reject(`Expected "toolName" to be a string but received ${typeof toolName} instead`);
    }
    dispatch({
      type: consts.SET_GL_FOR_TOOL,
      toolName,
      selectedGL
    });
  };
}

export function getProjectProgressForTools(toolName) {
  return (dispatch, getState) => {
    const {
      projectDetailsReducer: {
        projectSaveLocation,
        manifest
      }
    } = getState();
    const bookId = manifest.project.id;
    let progress = 0;
    if(typeof toolName !== 'string') {
      return Promise.reject(`Expected "toolName" to be a string but received ${typeof toolName} instead`);
    }
    const pathToCheckDataFiles = path.join(projectSaveLocation, INDEX_FOLDER_PATH, toolName, bookId);
    if (toolName === 'wordAlignment') {
      const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
      progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    } else {
      progress = ProjectDetailsHelpers.getToolProgress(pathToCheckDataFiles);
    }

    dispatch({
      type: consts.SET_PROJECT_PROGRESS_FOR_TOOL,
      toolName,
      progress
    });
  };
}

/**
 * @description Sends project manifest to the store
 * @param {object} manifest - manifest file of a project.
 * @return {object} action object.
 */
export function setProjectManifest(manifest) {
  return {
    type: consts.STORE_MANIFEST,
    manifest: manifest
  };
}

/**
 * @description adds a new key name to the manifest object
 * @param {String} propertyName - key string name.
 * ex.
 * manifest {
 *  ...,
 *  [propertyName]: 'value',
 *  ...
 * }
 * @param {*} value - value to be saved in the propertyName
 */
export function addObjectPropertyToManifest(propertyName, value) {
  return {
    type: consts.ADD_MANIFEST_PROPERTY,
    propertyName,
    value
  };
}

export function setProjectBookIdAndBookName() {
  return ((dispatch, getState) => {
    const { bookId } = getState().projectInformationCheckReducer;
    const bookName = bibleHelpers.convertToFullBookName(bookId);
    dispatch({
      type: consts.SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST,
      bookId,
      bookName
    });
  });
}

export function setProjectResourceId() {
  return ((dispatch, getState) => {
    const { resourceId } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_RESOURCE_ID_IN_MANIFEST,
      resourceId
    });
  });
}

export function setProjectNickname() {
  return ((dispatch, getState) => {
    const { nickname } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_NICKNAME_IN_MANIFEST,
      nickname
    });
  });
}

export function setLanguageDetails() {
  return ((dispatch, getState) => {
    const { languageDirection, languageId, languageName } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_LANGUAGE_DETAILS_IN_MANIFEST,
      languageDirection,
      languageId,
      languageName
    });
  });
}

export function updateContributors() {
  return ((dispatch, getState) => {
    const { contributors } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_TRANSLATORS_LIST_IN_MANIFEST,
      translators: contributors
    });
  });
}

export function updateCheckers() {
  return ((dispatch, getState) => {
    const { checkers } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_CHECKERS_LIST_IN_MANIFEST,
      checkers
    });
  });
}

/**
 * Change project name to match spec and handle overwrite warnings
 * @param {String} projectSaveLocation
 * @param {String} newProjectName
 * @return {Promise} - Returns a promise
 */
export function renameProject(projectSaveLocation, newProjectName) {
  return ((dispatch, getState) => {
    return new Promise(async (resolve) => {
      const projectPath = path.dirname(projectSaveLocation);
      const currentProjectName = path.basename(projectSaveLocation);
      const newProjectPath = path.join(projectPath, newProjectName);
      if (!fs.existsSync(newProjectPath)) {
        ProjectDetailsHelpers.updateProjectFolderName(newProjectName, projectPath, currentProjectName);
        dispatch(setSaveLocation(newProjectPath));
        resolve();
      }
      else { // project name already exists
        const translate = getTranslate(getState());
        const cancelText = translate('buttons.cancel_import_button');
        const continueText = translate('buttons.continue_import_button');
        dispatch(
          AlertModalActions.openOptionDialog(translate('projects.cancel_import_alert'),
            (result) => {
              if (result === cancelText) {
                // if 'cancel import' then close
                // alert and cancel import process.
                dispatch(AlertModalActions.closeAlertDialog());
                dispatch(cancelProjectValidationStepper());
              } else {
                // if 'Continue Import' then just close alert
                dispatch(AlertModalActions.closeAlertDialog());
              }
              resolve();
            },
            continueText,
            cancelText
          )
        );
      }
    });
  });
}

/**
 * handle rename prompting
 * @return {function(*, *): Promise<any>}
 */
export function doRenamePrompting() {
  return (async (dispatch, getState) => {
    const { projectDetailsReducer: {projectSaveLocation} } = getState();
    const hasGitRepo = fs.pathExistsSync(path.join(projectSaveLocation, '.git'));
    if (hasGitRepo) {
      // TODO: implement this in DCS renaming PR
      const newProjectName = path.basename(projectSaveLocation);
      dispatch(AlertModalActions.openOptionDialog("Your local project has been named\n  '" + newProjectName + "'.  \nPardon our mess, but DCS renaming to be implemented in future PR", () => {
        dispatch(AlertModalActions.closeAlertDialog());
      } ));
    } else { // no dcs
      dispatch(ProjectDetailsHelpers.showRenamedDialog());
    }
  });
}

/**
 * if project name needs to be updated to match spec, then project is renamed
 * @param {object} results of the renaming
 * @return {function(*, *): Promise<any>}
 */
export function updateProjectNameIfNecessary(results) {
  return (async (dispatch, getState) => {
    results.repoRenamed = false;
    results.repoExists = false;
    results.newRepoName = null;
    const {
      projectDetailsReducer: {manifest, projectSaveLocation}
    } = getState();
    const { repoNeedsRenaming, newRepoExists, newProjectName } = ProjectDetailsHelpers.shouldProjectNameBeUpdated(manifest, projectSaveLocation);
    results.newRepoName = newProjectName;
    if (repoNeedsRenaming) {
      if (newRepoExists) {
        results.repoExists = true;
      } else {
        results.repoRenamed = true;
        await dispatch(renameProject(projectSaveLocation, newProjectName));
      }
    }
  });
}

/**
 * if project name needs to be updated to match spec, then project is renamed
 * @return {function(*, *): Promise<any>}
 */
export function updateProjectNameIfNecessaryAndDoPrompting() {
  return (async (dispatch) => {
    const renamingResults = {};
    await dispatch(updateProjectNameIfNecessary(renamingResults));
    if (renamingResults.repoRenamed) {
      dispatch(doRenamePrompting());
    }
  });
}

/**
 * handles the prompting for overwrite/merge of project
 * @return {Promise} - Returns a promise
 */
export function handleOverwriteWarning(newProjectPath, projectName) {
  return ((dispatch, getState) => {
    return new Promise(async (resolve) => {
      const translate = getTranslate(getState());
      const confirmText = translate('buttons.overwrite_project');
      const cancelText = translate('buttons.cancel_import_button');
      let overwriteMessage = translate('projects.project_overwrite_has_alignment_message');
      if (! fs.existsSync(path.join(newProjectPath, '.apps'))) {
        overwriteMessage = (
          <div>
            <p>{translate('projects.project_already_exists', {'project_name': projectName})}</p>
            <p>{translate('projects.project_overwrite_no_alignment_message', {over_write: confirmText})}</p>
          </div>
        );
      }
      dispatch(
        AlertModalActions.openOptionDialog(overwriteMessage,
          (result) => {
            if (result === confirmText) {
              dispatch(AlertModalActions.closeAlertDialog());
              const oldProjectPath = path.join(PROJECTS_PATH, projectName);
              ProjectOverwriteHelpers.mergeOldProjectToNewProject(oldProjectPath, newProjectPath);
              fs.removeSync(oldProjectPath); // don't need the oldProjectPath any more now that .apps was merged in
              fs.move(newProjectPath, oldProjectPath); // replace it with new project
              resolve(true);
            } else { // if cancel
              dispatch(AlertModalActions.closeAlertDialog());
              resolve(false);
            }
          },
          cancelText,
          confirmText
        )
      );
    });
  });
}

export function updateProjectTargetLanguageBookFolderName() {
  return ((dispatch, getState) => {
    const {
      projectInformationCheckReducer: { bookId },
      projectDetailsReducer: { projectSaveLocation },
      localImportReducer: { oldSelectedProjectFileName }
    } = getState();
    if (!oldSelectedProjectFileName) {
      console.log("no old selected project File Name");
    } else {
      ProjectDetailsHelpers.updateProjectFolderName(bookId, projectSaveLocation, oldSelectedProjectFileName);
    }
  });
}
