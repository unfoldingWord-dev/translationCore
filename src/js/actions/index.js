/**
 * @author Ian Hoegen
 * @description: This serves as the new store and events api.
 * @usage:
 * Actions.sendAction('YOUR_EVENT', data) // Emits an event of type YOUR_EVENT, adds data to store
 * Actions.getData('YOUR_EVENT') // Retrieves data from the event emitted
 ***********************************************************************************/

const EventEmitter = require('events').EventEmitter;
const Reducers = require('../reducers');

class Actions extends EventEmitter {
  constructor() {
    super();
    this.store = [];
  }
  sendAction(eventName, actionObj) {
    if(Reducers[eventName]) {
      this.store[eventName] = Reducers[eventName](this.store[eventName], actionObj);
    } else {
      this.store[eventName] = actionObj;
    }
    this.emitChange(eventName);
  }

  subscribe(eventName, callback) {
    this.on(eventName, callback);
  }

  unsubscribe(eventName, callback) {
    this.removeListener(eventName, callback);
  }

  emitChange(eventName, data) {
    this.emit(eventName, data);
  }

  getData (actionType) {
    return this.store[actionType];
  }
}

var actions = new Actions();
module.exports = actions;
