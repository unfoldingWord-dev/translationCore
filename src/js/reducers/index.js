import { combineReducers } from 'redux';
// List of reducers
import { localizeReducer as localize } from 'react-localize-redux';
import toolsReducer from './toolsReducer';
import modalReducer from './modalReducer';
import loginReducer from './loginReducer';
import recentProjectsReducer from './recentProjectsReducer';
import importOnlineReducer from './importOnlineReducer';
import groupMenuReducer from './groupMenuReducer';// TODO
import settingsReducer from './settingsReducer';
import popoverReducer from './popoverReducer';
import resourcesReducer from './resourcesReducer';
import projectDetailsReducer from './projectDetailsReducer';
import alertModalReducer from './alertModalReducer';
// import commentsReducer from './commentsReducer';// TODO
import selectionsReducer from './selectionsReducer';// TODO
// import remindersReducer from './remindersReducer';// TODO
// import invalidatedReducer from './invalidatedReducer';// TODO
import contextIdReducer from './contextIdReducer';// TODO
import groupsDataReducer from './groupsDataReducer';// TODO
import groupsIndexReducer from './groupsIndexReducer';// TODO
// import verseEditReducer from './verseEditReducer';// TODO
import homeScreenReducer from './homeScreenReducer';
import myProjectsReducer from './myProjectsReducer';
import projectValidationReducer from './projectValidationReducer';
import copyrightCheckReducer from './copyrightCheckReducer';
import projectInformationCheckReducer from './projectInformationCheckReducer';
import mergeConflictReducer from './mergeConflictReducer';
import missingVersesReducer from './missingVersesReducer';
import localImportReducer from './localImportReducer';
import sourceContentUpdatesReducer from './sourceContentUpdatesReducer';
import softwareUpdateReducer from './softwareUpdateReducer';
import localeSettings from './localeSettings';
import alerts from './alerts';

// combining reducers
const rootReducers = combineReducers({
  localize,
  alerts,
  localeSettings,
  toolsReducer,
  modalReducer,
  loginReducer,
  settingsReducer,
  recentProjectsReducer,
  importOnlineReducer,
  groupMenuReducer,// TODO
  popoverReducer,
  resourcesReducer,
  projectDetailsReducer,
  alertModalReducer,
  // commentsReducer,// TODO
  selectionsReducer,// TODO
  // remindersReducer,// TODO
  // invalidatedReducer,
  contextIdReducer,// TODO
  groupsDataReducer,// TODO
  groupsIndexReducer,// TODO
  // verseEditReducer,// TODO
  homeScreenReducer,
  myProjectsReducer,
  projectValidationReducer,
  copyrightCheckReducer,
  projectInformationCheckReducer,
  mergeConflictReducer,
  missingVersesReducer,
  localImportReducer,
  sourceContentUpdatesReducer,
  softwareUpdateReducer,
});

export default rootReducers;
