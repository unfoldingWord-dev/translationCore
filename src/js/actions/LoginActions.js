import consts from './CoreActionConsts';
import * as CoreActions from './CoreActions.js';
import * as AlertModalActions from './AlertModalActions';
import gogs from '../components/core/login/GogsApi.js';
import { remote } from 'electron';
// const delclarations
const { dialog } = remote;
const api = window.ModuleApi;

export function setUserName(val) {
  return {
    type: consts.SET_USER_NAME,
    val: val
  }
}

export function setUserPassword(val) {
  return {
    type: consts.SET_USER_PASSWORD,
    val: val
  }
}

export function displayLogin(val) {
  return {
    type: consts.TOGGLE_ACOUNT_VIEW_TO_LOGIN,
    val: val
  }
}

export function loginLocalUser() {
  return ((dispatch, getState) => {
    let username = getState().loginReducer.userdata.username;
    dispatch({
      type: consts.LOGIN_LOCAL_USER,
      username
    });
  });
}

export function loginUser(newUserdata) {
  return ((dispatch) => {
    var Token = api.getAuthToken('gogs');
    gogs(Token).login(newUserdata).then((newUserdata) => {
      CoreActions.login(newUserdata);
      dispatch({
        type: consts.RECEIVE_LOGIN,
        val: newUserdata
      });
    }).catch(function(err) {
      var errmessage = "An error occurred while trying to login";
      if (err.syscall === "getaddrinfo") {
        errmessage = "Unable to connect to server";
      } else if (err.status === 404) {
        errmessage = "Incorrect Username";
      } else if (err.status === 401) {
        errmessage = "Incorrect Password";
      }
      dispatch(AlertModalActions.openAlertDialog(errmessage));
    });
  })
}

export function logoutUser(val) {
  return {
    type: consts.LOGOUT_USER
  }
}

export function feedbackChange(e) {
  return {
    type: consts.FEEDBACK_CHANGE,
    val: e
  }
}

export function subjectChange(subject) {
  return {
    type: 'FEEDBACK_SUBJECT_CHANGE',
    val: subject
  }
}

export function submitFeedback() {
  return {
    type: consts.SUBMIT_FEEDBACK
  }
}
