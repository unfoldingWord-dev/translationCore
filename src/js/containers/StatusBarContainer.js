import React from 'react';
import { connect  } from 'react-redux';
// Components
import StatusBar from '../components/core/SideBar/StatusBar.js';
// Actions
import * as modalActions from '../actions/ModalActions.js';
import * as coreStoreActions from '../actions/CoreActionsRedux.js';


class StatusBarContainer extends React.Component {

  render() {
    let { bookName } = this.props.projectDetailsReducer;
    let { toolName } = this.props.currentToolReducer;
    let { username } = this.props.loginReducer.userdata;

    return (
      <div>
        <StatusBar
          bookName={bookName}
          currentCheckNamespace={toolName}
          open={this.props.openModalAndSpecificTab}
          online={this.props.online}
          changeOnlineStatus={this.props.changeOnlineStatus}
          currentUser={username}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    statusBarReducer: state.statusBarReducer,
    currentToolReducer: state.currentToolReducer,
    projectDetailsReducer: state.projectDetailsReducer,
    loginReducer: state.loginReducer
  };
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    openModalAndSpecificTab: (tabkey, sectionKey, visible) => {
      dispatch(modalActions.selectModalTab(tabkey, sectionKey, visible));
    },
    changeOnlineStatus: val => {
      dispatch(coreStoreActions.changeOnlineStatus(val));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(StatusBarContainer);
