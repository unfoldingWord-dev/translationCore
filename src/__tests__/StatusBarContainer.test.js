/* eslint-env jest */

import os from 'os';
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import path from 'path-extra';
import Adapter from 'enzyme-adapter-react-16';
import { configure } from 'enzyme';
import StatusBarContainer from '../js/containers/StatusBarContainer';
import * as statusBarContainer from '../js/containers/StatusBarContainer';
import reducers from '../js/reducers';
import * as ProjectDetailsActions from '../js/actions/ProjectDetailsActions';
import * as BodyUIActions from '../js/actions/BodyUIActions';
import consts from '../js/actions/ActionTypes';
import * as LoginActions from '../js/actions/LoginActions';

const translate = key => key;

// TODO: this test should use a mock store

beforeAll(() => {
  configure({ adapter: new Adapter() });
});


// Tests for ProjectFAB React Component
describe('Test StatusBarContainer component',()=>{
  let store;

  beforeEach(() => {
    // create a new store instance for each test
    store = createStore(
      reducers,
      applyMiddleware(thunk)
    );
  });

  test('StatusBarContainer Component on current system should render button text correctly', () => {
    // given
    const projectName_ = 'en_tit_ulb';
    let projectFolder = '/user/dummy/tc/projects/';
    const projectPath = projectFolder + projectName_;
    const toolTitle = 'Miracle Tool';
    const username = 'Local User';
    setupStore(projectPath, toolTitle, username);
    // when
    const enzymeWrapper = (
      <Provider store={store}>
        <StatusBarContainer translate={translate}/>
      </Provider>
    );
    // then
    expect(enzymeWrapper).toMatchSnapshot();
  });

  test('StatusBarContainer Component on current system should match snapshot', () => {
    // given
    const osType = os.type().toLowerCase();
    const isWin = (osType.indexOf('win') === 0);
    const projectName_ = 'en_tit_ulb';
    let projectFolder = '/user/dummy/tc/projects/';

    if (isWin) { // if windows, switch to posix
      projectFolder = 'C:\\Users\\Dummy\\tC\\projects\\';
    }

    const projectPath = projectFolder + projectName_;
    const toolTitle = 'Miracle Tool';
    const username = 'Local User';
    setupStore(projectPath, toolTitle, username);

    // when
    const renderedValue = renderer.create(
      <Provider store={store}>
        <StatusBarContainer translate={translate}/>
      </Provider>
    ).toJSON();

    // then
    expect(renderedValue).toMatchSnapshot();
  });

  test('StatusBarContainer.getBaseName on mac/linux should render baseName correctly', () => {
    // given
    const expectedProjectName = 'en_tit_ulb';
    const projectPath = '/user/dummy/tc/projects/' + expectedProjectName;
    const posixPath = path.posix;

    // when
    const projectName = statusBarContainer.getBaseName(projectPath, posixPath);

    // then
    expect(projectName).toEqual(expectedProjectName);
  });

  test('StatusBarContainer.getBaseName on windows should render baseName correctly', () => {
    // given
    const expectedProjectName = 'en_tit_ulb';
    const projectPath = 'C:\\Users\\Dummy\\tC\\projects\\' + expectedProjectName;
    const winPath = path.win32;

    // when
    const projectName = statusBarContainer.getBaseName(projectPath, winPath);

    // then
    expect(projectName).toEqual(expectedProjectName);
  });

  //
  // Helpers
  //

  /**
   * @description initialize the store for testing
   * @param projectPath
   * @param toolTitle
   * @param username
   * @return {string}
   */
  function setupStore(projectPath, toolTitle, username) {
    store.dispatch(ProjectDetailsActions.setSaveLocation(projectPath));
    store.dispatch(BodyUIActions.toggleHomeView(false));
    // TODO: this will break. need to fix
    store.dispatch({
      type: consts.ADD_TOOL,
      name: 'default',
      tool: { title: toolTitle },
    });
    store.dispatch({
      type: consts.OPEN_TOOL,
      name: 'default',
    });

    const local = true;
    const userData = { username: username };
    store.dispatch(LoginActions.loginUser(userData, local));
  }
});

