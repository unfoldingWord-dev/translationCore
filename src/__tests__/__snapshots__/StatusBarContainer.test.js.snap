// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test StatusBarContainer component StatusBarContainer Component on current system should match snapshot 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "var(--background-color-dark)",
        "display": "flex",
        "flexDirection": "row",
        "height": "30px",
        "overflow": "hidden",
      }
    }
  >
    <div
      style={
        Object {
          "flexGrow": 1,
          "height": "100%",
          "overflow": "hidden",
        }
      }
    >
      <button
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "var(--background-color-dark)",
            "border": 0,
            "color": "var(--reverse-color)",
            "float": "left",
            "height": "100%",
            "minWidth": "200px",
            "outline": "none",
            "paddingLeft": 30,
            "paddingRight": 30,
            "width": "auto",
          }
        }
      >
        <span
          className="glyphicon glyphicon-home"
          style={
            Object {
              "fontSize": 15,
              "paddingRight": 8,
              "paddingTop": 3,
            }
          }
        />
        <span
          id="menu-text-root"
        >
          home
        </span>
      </button>
      <button
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "var(--background-color-dark)",
            "border": 0,
            "color": "var(--reverse-color)",
            "float": "left",
            "height": "100%",
            "minWidth": "200px",
            "outline": "none",
            "paddingLeft": 30,
            "paddingRight": 30,
            "width": "auto",
          }
        }
      >
        <span
          className="glyphicon glyphicon-user"
          style={
            Object {
              "fontSize": 15,
              "paddingRight": 5,
              "paddingTop": 3,
            }
          }
        />
        <span
          id="menu-text-root"
        >
          Local User
        </span>
      </button>
      <button
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "var(--background-color-dark)",
            "border": 0,
            "color": "var(--reverse-color)",
            "float": "left",
            "height": "100%",
            "minWidth": "200px",
            "outline": "none",
            "paddingLeft": 30,
            "paddingRight": 30,
            "width": "auto",
          }
        }
      >
        <span
          className="glyphicon glyphicon-folder-open"
          style={
            Object {
              "fontSize": 15,
              "paddingRight": 8,
              "paddingTop": 3,
            }
          }
        />
        <span
          id="menu-text-root"
        >
          en_tit_ulb
        </span>
      </button>
      <button
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "var(--background-color-dark)",
            "border": 0,
            "color": "var(--reverse-color)",
            "float": "left",
            "height": "100%",
            "minWidth": "200px",
            "outline": "none",
            "paddingLeft": 30,
            "paddingRight": 30,
            "width": "auto",
          }
        }
      >
        <span
          className="glyphicon glyphicon-wrench"
          style={
            Object {
              "float": "left",
              "fontSize": 15,
              "paddingRight": 5,
              "paddingTop": 3,
            }
          }
        />
        <div
          style={
            Object {
              "float": "left",
            }
          }
        >
          <span
            id="menu-text-root"
          >
            Miracle Tool
          </span>
        </div>
      </button>
    </div>
    <div
      style={
        Object {
          "margin": "auto 0",
          "padding": "0 50px",
        }
      }
    >
      <div>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`Test StatusBarContainer component StatusBarContainer Component on current system should render button text correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(StatusBarContainer)
    translate={[Function]}
  />
</Provider>
`;
